Add configurable extent buffer to symbols

JuhoErvasti opened the pull request:
Fixes #38579.

Allows the user to set an extent buffer per symbol. Useful with for example geometry generator symbols:

https://github.com/user-attachments/assets/2a81d593-0817-4a6e-9be0-0f8b7a4e3852

You can set the value in different units and it can also be controlled by an expression (for example you can set a different value based on the map scale).

When rendering `QgsRenderer` finds the biggest buffer value that has been set in all of its symbols. The extent for the feature request in `QgsVectorLayerRenderer` is then modified according to that value.

Backport to 3.40 requested (if possible).

Funded by the National Land Survey of Finland.

nyalldawson commented:
@JuhoErvasti thanks for tackling this annoying problem!

I have a few thoughts on the approach used here:

- I don't think this needs to be a symbol-layer property, rather it makes more sense to be a setting on a whole symbol itself. What's the pressing use case where one symbol layer would need a different size to the rest of the symbol? In my opinion doing this at the symbol layer considerably complicates both the code + the user experience.
- If we made this a symbol setting, then the setting could be moved to sit inside the symbol "advanced" menu. This would help avoid the UI clutter that results from adding a rarely-changed, advanced setting directly inside ALL symbol layer property windows. (There's also a side issue caused by adding this spin box in the horizontal line with the existing widgets, as it considerably increases the minimum width for the panel, which will cause issues on smaller screen sizes)
- What's the use case for data-definable buffers? I can't see how we'd ever need to set this on a per-feature basis. And right now this could ONLY be used to decrease the buffer size from the fixed amount, which sounds like an incredibly esoteric requirement. If there's a need to set data-definable buffer size eg to handle logic with different map scales/crs/..., then it should be evaluated ONCE for the layer and not on a per-feature basis (and make sure that the widget for it doesn't expose feature related context variables)

JuhoErvasti commented:
@nyalldawson
> I don't think this needs to be a symbol-layer property

I agree that making it a symbol layer property does complicate things and the use cases for separate buffer levels are probably few and far between. The reason I did it this way was based on your [comment](https://github.com/qgis/QGIS/issues/38579#issuecomment-694414693) from the issue's discussion:
> I'm +1 for a configurable buffer parameter, but it needs to be done at the symbol layer level.

I interpreted this to mean that it does need to be a symbol layer property, but reading the thread again you might've meant it as a symbol property (as opposed to layer/project property as suggested by others)? Either way if that isn't the requirement, I'm happy to change it to be on the symbol level.

> If we made this a symbol setting, then the setting could be moved to sit inside the symbol "advanced" menu.

Sounds good.

> What's the use case for data-definable buffers?

I figured it'd be good to be able to control the buffer with an expression (for example to make it scale-dependant as you mentioned) and since I was adding it as a symbol layer property it made sense to use a data-defined button since almost everything else has one. But if the location of the property is changed, this can of course be changed too.

Could it be even just a normal expression widget, not a data-defined override button?

> If there's a need to set data-definable buffer size eg to handle logic with different map scales/crs/..., then it should be evaluated ONCE for the layer

If it's done on the symbol level, how should this work? If you're using f.e. the graduated renderer you could set a different expression/value for each symbol, same with the geometry generator subsymbols. Wouldn't it have to be evaluated at least once per each symbol in the renderer?

nyalldawson commented:
@JuhoErvasti 
> I agree that making it a symbol layer property does complicate things and the use cases for separate buffer levels are probably few and far between. The reason I did it this way was based on your https://github.com/qgis/QGIS/issues/38579#issuecomment-694414693 from the issue's discussion:

Sorry, must have been an accidental mis-phrasing at the time. I think I was referring to "should be at the symbol level" as opposed to "should be one setting at the project (or layer) level", but accidentally wrote "symbol layer". 

> I interpreted this to mean that it does need to be a symbol layer property, but reading the thread again you might've meant it as a symbol property (as opposed to layer/project property as suggested by others)? Either way if that isn't the requirement, I'm happy to change it to be on the symbol level.

> I figured it'd be good to be able to control the buffer with an expression (for example to make it scale-dependant as you mentioned) and since I was adding it as a symbol layer property it made sense to use a data-defined button since almost everything else has one. But if the location of the property is changed, this can of course be changed too.

> Could it be even just a normal expression widget, not a data-defined override button?

Elsewhere we do use data-defined buttons for layer-level rendering properties. Eg for raster layer opacity you can data-define, even though it's only evaluated once just before the layer render. I'd say this could be similar -- we could evaluate once upfront at the start of the vector layer renderer, and then just use that to control the feature request extent. Nice and simple and we then avoid ALL handling when doing per-symbol-rendering.


> If it's done on the symbol level, how should this work? If you're using f.e. the graduated renderer you could set a different expression/value for each symbol, same with the geometry generator subsymbols. Wouldn't it have to be evaluated at least once per each symbol in the renderer?

For multi-symbol renderers I'd do it once per symbol upfront, and then take the max and that's the value used for the feature request. The worst case scenario there is that we end up rendering a few extra symbols which fall totally outside the map.

JuhoErvasti commented:
@nyalldawson

If I could ask for one more clarification:
>I'd say this could be similar -- we could evaluate once upfront at the start of the vector layer renderer, and then just use that to control the feature request extent. Nice and simple and we then avoid ALL handling when doing per-symbol-rendering.

>For multi-symbol renderers I'd do it once per symbol upfront, and then take the max and that's the value used for the feature request.

I might be missing something, but if the extent buffer should only be applied to the feature request and there shouldn't be any further handling what is the benefit of having separate values for different symbols? Doesn't that make it effectively a renderer-level property even if the values are set for the symbols?

nyalldawson commented:
@JuhoErvasti good question!

My thinking is that this property is innately associated with the shape and size of a particular symbol, so should be attached to that. That way, the property will be stored inside the symbol if you eg copy it to another layer or save to the style library and reuse the symbol in a different layer, and won't require re-configuring the renderer setting each time.

JuhoErvasti committed:
Add extent buffer property to QgsSymbol

JuhoErvasti committed:
Add maximumExtentBuffer() function to QgsFeatureRenderer

JuhoErvasti committed:
Add tests for maximumExtentBuffer()

JuhoErvasti committed:
Modify feature request extent based on max extent buffer

JuhoErvasti committed:
Add tests for rendering with extent buffer

nyalldawson commented on src/core/symbology/qgsrenderer.h:
```diff
@@ -653,6 +653,15 @@ class CORE_EXPORT QgsFeatureRenderer
     void copyRendererData( QgsFeatureRenderer *destRenderer ) const;
 
+    /**
+     * Returns the maximum extent buffer found in this renderer's
+     * symbols' symbol layers.
```
```suggestion
     * Returns the maximum extent buffer found in this renderer's symbols.
```

nyalldawson commented on python/PyQt6/gui/auto_generated/symbology/qgsextentbufferdialog.sip.in:
```diff
@@ -0,0 +1,117 @@
+
+
+
+class QgsExtentBufferWidget : QgsPanelWidget, QgsExpressionContextGenerator
```
Let's SIP_SKIP this file, there's no need to expose this widget to Python

nyalldawson commented on src/core/symbology/qgssymbol.cpp:
```diff
@@ -20,12 +20,14 @@
 #include <QSvgGenerator>
 #include <QPicture>
 
+#include <algorithm>
```
```suggestion
```

nyalldawson commented on src/core/symbology/qgssymbollayerutils.cpp:
```diff
@@ -1452,6 +1453,7 @@ QDomElement QgsSymbolLayerUtils::saveSymbol( const QString &name, const QgsSymbo
   symEl.setAttribute( QStringLiteral( "name" ), name );
   symEl.setAttribute( QStringLiteral( "alpha" ), QString::number( symbol->opacity() ) );
   symEl.setAttribute( QStringLiteral( "clip_to_extent" ), symbol->clipFeaturesToExtent() ? QStringLiteral( "1" ) : QStringLiteral( "0" ) );
+  symEl.setAttribute( QStringLiteral( "extent_buffer" ), QString::number( symbol->extentBuffer() ) );
```
Let's keep the XML as small as necessary:
```suggestion
  if ( !qgsDoubleNear( symbol->extentBuffer(), 0 ) )
    symEl.setAttribute( QStringLiteral( "extent_buffer" ), QString::number( symbol->extentBuffer() ) );
```

nyalldawson commented on src/core/symbology/qgsrenderer.cpp:
```diff
@@ -417,6 +418,38 @@ QgsLegendSymbolList QgsFeatureRenderer::legendSymbolItems() const
+  {
+    const QgsProperty property = sym->dataDefinedProperties().property( QgsSymbol::Property::ExtentBuffer );
+
+    if ( property.isActive() && ! property.expressionString().isEmpty() )
```
```suggestion
    if ( property.isActive() )
```

nyalldawson commented on src/core/symbology/qgsrenderer.cpp:
```diff
@@ -417,6 +418,38 @@ QgsLegendSymbolList QgsFeatureRenderer::legendSymbolItems() const
+
+    if ( property.isActive() && ! property.expressionString().isEmpty() )
+    {
+      return sym->dataDefinedProperties().valueAsDouble( QgsSymbol::Property::ExtentBuffer, expContext );
```
```suggestion
      return sym->dataDefinedProperties().valueAsDouble( QgsSymbol::Property::ExtentBuffer, expContext, sym->extentBuffer() );
```

nyalldawson commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -354,8 +355,26 @@ bool QgsVectorLayerRenderer::renderInternal( QgsFeatureRenderer *renderer, int r
+
   renderer->modifyRequestExtent( requestExtent, context );
 
+  double maximumExtentBuffer = renderer->maximumExtentBuffer( context );
+
+  if ( maximumExtentBuffer != 0 )
+  {
+    bool bufferDisappearsExtent = maximumExtentBuffer < 0 && ( requestExtent.width() + ( maximumExtentBuffer * 2 ) < 0 || requestExtent.height() + ( maximumExtentBuffer * 2 ) < 0 );
+
+    // nothing to draw
+    if ( bufferDisappearsExtent )
+    {
+      renderer->stopRender( context );
+      return true;
+    }
+
+    requestExtent = requestExtent.buffered( maximumExtentBuffer );
+  }
```
It might be cleaner to just move this logic into  QgsFeatureRenderer::modifyRequestExtent ?

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -867,6 +868,23 @@ class CORE_EXPORT QgsSymbol
+
+    /**
+     * Sets the symbol's extent buffer.
+     *
```
This needs some comment about permitting a negative buffer size, which would cause the symbol not to be rendered if it's close to the edge of the map.

But that said -- negative sizes won't work well for anything but single symbol renderer, because if you have any symbol in the other renderers with the default 0 size then maximumExtentBuffer will ignore the negative values and just return 0...

Are negative buffers necessary? I'd drop them if not

nyalldawson commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -354,8 +355,26 @@ bool QgsVectorLayerRenderer::renderInternal( QgsFeatureRenderer *renderer, int r
+
+  if ( maximumExtentBuffer != 0 )
+  {
+    bool bufferDisappearsExtent = maximumExtentBuffer < 0 && ( requestExtent.width() + ( maximumExtentBuffer * 2 ) < 0 || requestExtent.height() + ( maximumExtentBuffer * 2 ) < 0 );
```
```suggestion
    bool bufferCausesEmptyExtent = maximumExtentBuffer < 0 && ( requestExtent.width() + ( maximumExtentBuffer * 2 ) < 0 || requestExtent.height() + ( maximumExtentBuffer * 2 ) < 0 );
```

nyalldawson commented on src/gui/symbology/qgsextentbufferdialog.cpp:
```diff
@@ -0,0 +1,128 @@
+
+QgsExpressionContext QgsExtentBufferWidget::createExpressionContext() const
+{
+  QList<QgsExpressionContextScope *> scopes = mContext.globalProjectAtlasMapLayerScopes( mLayer );
```
You'll also need to handle QgsSymbolWidgetContext if you want access to the canvas in order to get the @map_scale variable

JuhoErvasti replied:
@nyalldawson 

Is it not sufficiently handled? When I'm testing I can use the @map_scale etc. variables. `mContext` here is a `QgsSymbolWidgetContext` object and it is set for the widget when it is created in `QgsSymbolListWidget`. Is there something else that needs to be done?

nyalldawson replied:
Have a look at eg how QgsTextFormatWidget::createExpressionContext handles this -- the logic here should be almost identical

nyalldawson commented on src/gui/symbology/qgssymbolslistwidget.cpp:
```diff
@@ -59,6 +60,9 @@ QgsSymbolsListWidget::QgsSymbolsListWidget( QgsSymbol *symbol, QgsStyle *style,
   mAnimationSettingsAction = new QAction( tr( "Animation Settings…" ), this );
   connect( mAnimationSettingsAction, &QAction::triggered, this, &QgsSymbolsListWidget::showAnimationSettings );
 
+  mExtentBufferAction = new QAction( tr( "Extent buffer…" ), this );
```
```suggestion
  mExtentBufferAction = new QAction( tr( "Extent Buffer…" ), this );
```

nyalldawson commented:
Thanks @JuhoErvasti !

One other consideration raised by @nirvn -- shouldn't we also add a unit selection for this setting? It would be more portable between different layers if the user could set the buffer size in eg millimeters instead of just map units (which may differ in size from layer to layer). 

QgsRenderContext.convertToMapUnits could be used to convert the size + unit to a map unit size when required.

JuhoErvasti committed:
Add test for loading symbol with extent buffer

JuhoErvasti committed:
Fix style

JuhoErvasti committed:
Add a way to change extent buffer in the GUI

JuhoErvasti committed:
Apply suggestions from review

JuhoErvasti committed:
Disallow negative buffer

JuhoErvasti committed:
Add unit selection for extent buffer

JuhoErvasti committed:
Check extent buffer size unit in tests

JuhoErvasti commented:
@nyalldawson 

Thank you for the review! I've made the changes you suggested. I ended up dropping negative buffers since I didn't really have a clear use case for the in mind. I originally thought that I might as well allow it in case someone has an obscure use case for them but as you pointed out now that the extent buffer isn't in the symbol layer it wouldn't work that well.

I also added the option to change the size unit.

nyalldawson commented on src/core/symbology/qgssymbol.cpp:
```diff
@@ -2239,6 +2254,7 @@ void QgsSymbol::initPropertyDefinitions()
   sPropertyDefinitions = QgsPropertiesDefinition
   {
     { static_cast< int >( QgsSymbol::Property::Opacity ), QgsPropertyDefinition( "alpha", QObject::tr( "Opacity" ), QgsPropertyDefinition::Opacity, origin )},
+    { static_cast< int >( QgsSymbol::Property::ExtentBuffer ), QgsPropertyDefinition( "extent_buffer", QObject::tr( "Extent buffer" ), QgsPropertyDefinition::Double, origin )},
```
```suggestion
    { static_cast< int >( QgsSymbol::Property::ExtentBuffer ), QgsPropertyDefinition( "extent_buffer", QObject::tr( "Extent buffer" ), QgsPropertyDefinition::DoublePositive, origin )},
```

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -867,6 +868,38 @@ class CORE_EXPORT QgsSymbol
+    /**
+     * Returns the symbol's extent buffer.
+     *
+     * \returns The symbol's extent buffer
```
```suggestion
     * Units are retrieved via extentBufferSizeUnit().
```

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -867,6 +868,38 @@ class CORE_EXPORT QgsSymbol
+
+    /**
+     * Sets the symbol's extent buffer.
+     *
```
```suggestion
     *
     * Units are set via setExtentBufferSizeUnit().
```

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -867,6 +868,38 @@ class CORE_EXPORT QgsSymbol
+    /**
+     * Returns the units for the buffer size.
+     *
+     * \see setExtentBufferSizeUnit()
```
```suggestion
     * \see extentBuffer()
     * \see setExtentBufferSizeUnit()
     *
     * \since QGIS 3.42
```

nyalldawson commented on src/core/symbology/qgssymbol.h:
```diff
@@ -867,6 +868,38 @@ class CORE_EXPORT QgsSymbol
+    /**
+     * Sets the \a unit used for the extent buffer.
+     *
+     * \see extentBufferSizeUnit()
```
```suggestion
     * \see setExtentBuffer()
     * \see extentBufferSizeUnit()
     *
     * \since QGIS 3.42
```

nyalldawson commented on src/core/vector/qgsvectorlayerrenderer.cpp:
```diff
@@ -17,6 +17,7 @@
 
 #include "qgsmessagelog.h"
 #include "qgspallabeling.h"
+#include "qgsrectangle.h"
```
```suggestion
```

nyalldawson commented on src/gui/symbology/qgssymbolslistwidget.cpp:
```diff
@@ -299,6 +304,47 @@ void QgsSymbolsListWidget::showAnimationSettings()
+  }
+  else
+  {
+    QgsExtentBufferDialog dlg( mSymbol, mLayer, panel );
```
We should also have a "setContext" call here for the QgsExtentBufferDialog class, which in turn sets it for the QgsExtentBufferWidget. That'll give us consistency if this dialog is opened via the layer properties window.

nyalldawson commented:
Thanks @JuhoErvasti -- a few more minor comments here

JuhoErvasti committed:
Address review

JuhoErvasti commented:
@nyalldawson 

Thanks, I've addressed the points. Hopefully everything's OK now. Also I realized that the data defined property still could've produced a negative buffer so that should be handled now, as well as handling the `@value` variable

JuhoErvasti committed:
Merge remote-tracking branch 'upstream/master' into fix_38579_symbol_rendering_out_of_extent

JuhoErvasti committed:
Fix format

nyalldawson commented:
@JuhoErvasti there's one bit missed in [7089713](https://github.com/qgis/QGIS/pull/59630/commits/7089713eb9ed05cb956dbe68b8791016283d98c8) -- we still need the \since added to extentBufferSizeUnit / setExtentBufferSizeUnit

JuhoErvasti committed:
Add \since to new functions

JuhoErvasti commented:
@nyalldawson 

Sorry, added now!

nyalldawson commented:
Great, thanks for your hard work @JuhoErvasti !

DelazJ commented:
@JuhoErvasti thanks for fixing that issue. Could you please update your top message to reflect the changes before we tag it for docs and changelog?

JuhoErvasti commented:
@DelazJ 

I've updated the message now.
